[dart 3.9.0] Hooks guide (#6759)
dart-lang/site-www: src/content/tools/hooks.md

description: Run custom build scripts.
---

:::note
Support for build hooks is in **preview** and
can be used on the `main` and `beta` channels.
:::

This guide explains what hooks are and how to use them
with a package.

## Introduction

Hooks are Dart scripts placed in the `hook/` directory of
your Dart package. These scripts have a predefined format
for their input and output. With hooks, the Dart SDK can:

1. Discover your hooks.
1. Execute your hooks with the necessary input.
1. Consume the output produced by the hooks.

Example project:

```plaintext
📁  example_project                   // project with hooks
    📁  hook                          // add hook scripts here
        📄 build.dart
    📁  lib                           // reference your assets here
        📄 example.dart
    📁  src                           // add native sources here
        📄 example_native_library.c
        📄 example_native_library.h
    📁  test                          // test your assets here
        📄 example_test.dart
```

## Hooks

Currently, the build hook is available, but more are
planned. To learn more, see the following.

### Build hooks {:.no_toc}

A build hook builds assets that you want

> to run during your project's build process. This is useful

Author: Yes, this is confusing. I've updated this section in my next revision. Please take a close look at my changes.

for tasks that need to happen before your main
application code is compiled, such as:

* Compiling C, C++, Rust, or other native code into a
  dynamic library.
* Downloading or generating other assets required at
  runtime.

A package's build hook is placed in the `hook/build.dart`

> directory. Use the [`build`][] function to parse the hook

Author: Added and made some updates in this section.

Therefore, build hooks are run in the order of dependencies.

[`assets`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput/assets.html
[`build`]: {{site.pub-api}}/hooks/latest/hooks/build.html
[`BuildInput`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput-class.html
[`BuildOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/BuildOutputBuilder-class.html
[`BuildInput.sharedOutputDirectory`]: {{site.pub-api}}/hooks/latest/hooks/HookInput/outputDirectoryShared.html
[`metadata`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput/metadata.html

## Assets

Assets are the files output from a hook that need to be
bundled in a Dart application. Assets can be accessed
at runtime from the Dart code. Currently, the Dart SDK can
use the `CodeAsset` type, but more are planned.
To learn more, see the following.

### CodeAsset type {:.no_toc}

A [`CodeAsset`][] represents a code asset. A code asset is
a dynamic library compiled from a language other than
Dart, such as C, C++, Rust, or Go. `CodeAsset` is part of
the `code_asset` package and are called with `dart:ffi` from
the Dart code at runtime using the `@Native` annotation.

> A code asset can be used with build and link hooks.

Reviewer: Should we remove the referene to link hook here since we explain nowhere what that is yet?

`pubspec.yaml`, and then add the following dependencies as
needed under `dependencies`:

* `hooks`
* `code_assets`

You will also likely need to add a tool for your specific
task, such as `native_toolchain_c` for compiling C code.

:::note
You need to add the dependencies under `dependencies`, not
`dev_dependencies`. The hooks will be run by packages and
applications depending on your package, so the
Dart code needs to be part of the resolution in those
packages.
:::

Example dependencies for a build hook:

```yaml title="pubspec.yaml"
name: native_add_library
description: Sums two numbers with native code.
version: 0.1.0

environment:
  sdk: '>=3.9.0 <4.0.0'

dependencies:
  ... 
  code_assets: any
  hooks: any
  native_toolchain_c: any

dev_dependencies:
  ...
  ffigen: ^18.0.0
 ```

### Create a build hook script {: #create-hook }

If you want to use a build hook, first create the native
library that you want to use, and then follow these steps to
connect a build hook to that library in a Dart script: 

1.  In your Dart project, create or open `hooks/build.dart`. 

1.  In the main method, add a `build` function from
    `package:hooks/hooks.dart` and link to the native

>     library. For example:

Reviewer: instead of "and link to the native library" maybe: "... and use the appropriate toolchain to compile the native library."

    ```dart title="hooks/build.dart" highlightLines=6
    import 'package:hooks/hooks.dart';
    import 'package:logging/logging.dart';
    import 'package:native_toolchain_c/native_toolchain_c.dart';

    void main(List<String> args) async {
      await build(args, (input, output) async {
        final packageName = input.packageName;
        final cbuilder = CBuilder.library(
          name: packageName,
          assetName: '$packageName.dart',
          sources: ['src/$packageName.c'],
        );
        await cbuilder.run(
          input: input,
          output: output,
          logger: Logger('')
            ..level = Level.ALL
            ..onRecord.listen((record) => print(record.message)),
        );
      });
    }
    ```

> 

Reviewer: Maybe this could get an introductory sentence like: Build hooks are given two arguments:

>     * `input`: The input for `hook/build.dart`. This

Reviewer: Maybe it would be useful to include that the `input` is read-only and it includes information for the hook to produce the right asset type, e.g. like target OS etc.

Author: Yes, definitely. Updated!

      includes the target OS, target architecture,
      output directory, etc. For details, see
      the [`BuildInput`][] class.

>     * `output`: The output for `hook/build.dart`. This

Reviewer: Maybe include here that this is a builder object and it is the responsibility of the build hook to set the appropriate information on it. What asset have been produced? etc.

In other words (and maybe we can include this somewhere in the intro sentence above to connect it all together): The build hook script is supposed to read from the inputs, produce an asset, and then fill the output with what it produced.

Author: Thanks for the clarification. Please take a close look at my changes.

      includes the assets to be bundled, and dependencies.
      For details, see the [`BuildOutputBuilder`][] class.

[`BuildInput`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput-class.html
[`BuildOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/BuildOutputBuilder-class.html

### Generate or download assets {: #generate-assets }

> Assets are generated and/or downloaded when you invoke your

Reviewer: Since you never really invoke a hook directly, maybe better: Assets are generated and/or downloaded according to your hook script when your hook is invoked by the Dart SDK, which autoamtically happens when you use the `run`, `build`, or `test` command on your package.

Dart hooks. Your Dart hooks are invoked when you use the
`run`, `build`, or `test` command on your package.
The resulting assets are stored in the directory that you

> specified with [`BuildInput.sharedOutputDirectory`][]. 

Reviewer: I am not sure what this sentence is supposed to tell me. It sounds like it wants to tell me how to set the output directory for an asset, but it then links to`BuildInput.sharedOutputDirectory`, which is a getter.

There's also kind of a disconnect to the example given above, which doesn't seem to specify an output directory at all (it probably happens indirectly inside the CBuilder). Can we somehow connect this back to the example so people are not confused?

[`BuildInput.sharedOutputDirectory`]: {{site.pub-api}}/hooks/latest/hooks/HookInput/outputDirectoryShared.html

### Reference assets {: #reference-assets }

Assets are the files that hooks create. Once an asset has been created,
you can reference it in your code and at runtime with it's

> asset ID ([`assetId`][]).

Reviewer: I feel like there is kind of a disconnect between this sentence and the example below because the example below never refers to an asset id.

Reviewer 2: I've reworked it.

I kept all the asset-id stuff in the reference-asset section, because you don't
really think about it during the build-time. The alternative would be to
already explain it in the create-hook section, saying that it will be important
later. WDYT?

Reviewer: I think this is great now. Thanks for the rewrite!

The following example illustrates how to call a native
C function `add` from `native_add_library.c`:

```dart title="my_package/lib/my_package.dart"
import 'dart:ffi' as ffi;

@ffi.Native<ffi.Int32 Function(ffi.Int32, ffi.Int32)>()
external int add(int a, int b);
```

:::note
*   The asset ID is constructed in the build hook in the example above as
    `package:$packageName/$packageName.dart` (which is
    `package:native_add_library/native_add_library.dart`).
*   The asset ID in `@Native` is optional and defaults to
    the library uri (which also is
    `package:native_add_library/native_add_library.dart` in the example above).
:::

[`assetId`]: {{site.dart-api}}/dart-ffi/Native/assetId.html

### Test assets {: #test-assets }

After you've generated an asset and referenced it in
your Dart code, consider creating a test for the generated

> asset. In the following example, a test is created for

Author: Sounds good. Added updates.

`native_add_library.dart`, a script that references a

> code asset called  

Author: Fixed!

        final cbuilder = CBuilder.library(
          name: packageName,
          assetName: '$packageName.dart',
          sources: ['src/$packageName.c'],
        );
        await cbuilder.run(
          input: input,
          output: output,
          logger: Logger('')
            ..level = Level.ALL
            ..onRecord.listen((record) => print(record.message)),
        );
      });
    }
    ```

    * `input`: The input for `hook/build.dart`. This
      includes the target OS, target architecture,
      output directory, etc. For details, see
      the [`BuildInput` class][].

    * `output`: The output for `hook/build.dart`. This
      includes the assets to be bundled, and dependencies.
      For details, see the [`BuildOutputBuilder` class][].

### Create a link hook script

If you want to use a link hook, follow these steps to
add one to a Dart script: 

1.  Open the script that you want to contain the build hook
    (for example, `hooks/link.dart`). 

1.  In the main method, add a `build` function from
    `package:hooks/hooks.dart`. For example:

    ```dart title="hooks/link.dart"
    import 'dart:convert';
    import 'dart:io';

>     import 'package:data_assets/data_assets.dart';

Author: I don't think we support this yet, correct?

Reviewer 2: Correct.

(Long answer: The data_assets package can be used in the hook during builds. But
the data assets cannot yet be accessed at runtime by the Dart code.)

        await cbuilder.run(
          input: input,
          output: output,
          logger: Logger('')
            ..level = Level.ALL
            ..onRecord.listen((record) => print(record.message)),
        );
      });
    }
    ```

    The `build` function takes two arguments:

    * `input`: (readonly) The input for the hook.
      Includes information for the hook
      to produce the right asset type (for example,
      target OS, target architecture, output directory, and
      more). For details, see the [`BuildInput`][] class.

    * `output`: The output for the hook. After the
      build hook reads the input, it produces an asset and
      then provides what it produced as the output.
      For details, see the [`BuildOutputBuilder`][] class.

[`BuildInput`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput-class.html
[`BuildOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/BuildOutputBuilder-class.html

### Automatically add assets {: #generate-assets }

Assets are added indirectly when you use the `run`,
`build`, or `test` command to invoke hooks on your package.
The resulting assets are stored in the directory that you
specified in your hooks script. 

> ### Reference assets {: #reference-assets }

Reviewer 2: Maybe this should be "Use assets"

Basically the build step happens at "build time" and using an asset happens at
"run time". Would standardizing on build-time and runtime in this doc clarify
things?

(Using "create" as the verb for assets at build-time is fine. Create and use.
Create and reference.)

Reviewer: ```suggestion
### Reference assets from Dart code {: #reference-assets }
```

Author: I think `Use assets` will work the best since it's brief, looks good in the TOC, and is easy to parse quickly. @dcharkes and @goderbauer fyi.

You can currently use hooks to do things
such as compile or download native assets (code written in
other languages that are compiled into machine code), and
then call these assets from the Dart code of a package.

Hooks are Dart scripts placed in the `hook/` directory of
your Dart package. They have a predefined format for their
input and output, which allows the Dart SDK to:

1. Discover the hooks.
1. Execute the hooks with the necessary input.
1. Consume the output produced by the hooks.

Example project with a build hook:

```plaintext
📁  example_project                   // project with hooks
    📁  hook                          // add hook scripts here
        📄 build.dart
    📁  lib                           // reference your assets here
        📄 example.dart
    📁  src                           // add native sources here
        📄 example_native_library.c
        📄 example_native_library.h
    📁  test                          // test your assets here
        📄 example_test.dart
```

## Hooks

Currently, the build hook is available, but more are
planned. To learn more, see the following.

### Build hooks {:.no_toc}

With build hooks, a package can do things such as

> compile or download native assets such as C or Rust libraries.

Author: Added the entry to this list for now: https://github.com/dart-lang/site-www/issues/6461

Afterwards, these assets can be called from the Dart code of
a package.

> A package's build hook is automatically invoked by the

Reviewer 3: Can a package only have one build hook?

Dart SDK at an appropriate time during the build process.
Build hooks are run in parallel with Dart compilation and
might do longer running operations such as downloading or
calling a native compiler.

Use the [`build`][] function to parse the hook
input with [`BuildInput`] and then write the hook output
with [`BuildOutputBuilder`]. The hook should place
downloaded and generated assets in
[`BuildInput.sharedOutputDirectory`][].

The assets produced for your package might depend on
[`assets`][] or [`metadata`][] produced by the build hooks
from the packages in your direct dependencies. Therefore,

> build hooks are run in the order of dependencies.

Reviewer 3: The order of what dependencies? From a package's pubspec file, declared in the hook output?

Reviewer 2: Good catch! The pubspec dependencies.

[`BuildInput.sharedOutputDirectory`]: {{site.pub-api}}/hooks/latest/hooks/HookInput/outputDirectoryShared.html
[`metadata`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput/metadata.html

## Assets

Assets are the files that are produced by a hook and then
bundled in a Dart application. Assets can be accessed
at runtime from the Dart code. Currently, the Dart SDK can
use the `CodeAsset` type, but more asset types are planned.
To learn more, see the following.

### CodeAsset type {:.no_toc}

A [`CodeAsset`][] represents a code asset. A code asset is
a dynamic library compiled from a language other than
Dart, such as C, C++, Rust, or Go. `CodeAsset` is part of
the `code_asset` package and are called with `dart:ffi` from
the Dart code at runtime using the `@Native` annotation.

[`CodeAsset`]: {{site.pub-api}}/code_assets/latest/code_assets/CodeAsset-class.html

## Use a hook {: #use-hooks-assets }

To add assets to your project, use a hook. For details,
see the following sections.

### Add dependencies {: #add-dependencies-hooks-assets }

To use a hook, you must first add it to your `pubspec.yaml`

> dependencies.

Reviewer 3: Unless there's a specific reason developers need to add these dependencies to the file manually, do suggest they use `dart pub add hooks code_assets` instead. It's what we've been moving to elsewhere and is less prone to mistakes.

1. Open your package in a code editor.

1. Navigate to `pubspec.yaml`.

1. Go to the `dependencies` section and add the following:

    * `hooks`
    * `code_assets`

1. To this list, also add any tools for your specific task,
   such as `native_toolchain_c` for compiling C code.

:::note
You need to add the dependencies under `dependencies`, not
`dev_dependencies`. The hooks will be run by packages and
applications depending on your package, so the
Dart code needs to be part of the resolution in those
packages.
:::

Example dependencies for a build hook:

```yaml title="pubspec.yaml"
name: native_add_library
description: Sums two numbers with native code.
version: 0.1.0

environment:

>   sdk: '>=3.9.0 <4.0.0'

Reviewer 3: ```suggestion
  sdk: ^3.9.0
```

dependencies:

>   ... 

Reviewer 3: So that this file can be valid yaml, in case it's ever copy pasted.

```suggestion
  # ... 
```

  code_assets: any
  hooks: any
  native_toolchain_c: any

dev_dependencies:

>   ...

Reviewer 3: ```suggestion
  # ...
```

  ffigen: ^18.0.0
 ```

### Create a build hook to generate native assets {: #create-hook }

If you want to use a build hook to transparently compile
native assets (such as C or Rust libraries), which are then
made available to be called from the Dart code of a package,
create a `build.dart` script similar to the following:

1.  In your Dart project, create or open `hooks/build.dart`. 

> 1.  In the `main` method, add a `build` function from
>     `package:hooks/hooks.dart` and use the appropriate

Reviewer 3: ```suggestion
1.  In the `main` method, call the `build` function from
    `package:hooks/hooks.dart` and use the appropriate
```

    toolchain to compile the native library. For example:

    ```dart title="hooks/build.dart" highlightLines=6
    import 'package:hooks/hooks.dart';
    import 'package:logging/logging.dart';
    import 'package:native_toolchain_c/native_toolchain_c.dart';

    void main(List<String> args) async {
      await build(args, (input, output) async {
        final packageName = input.packageName;

>         final cbuilder = CBuilder.library(
>           name: packageName,
>           assetName: '$packageName.dart',
>           sources: ['src/$packageName.c'],
>         );
>         await cbuilder.run(

Reviewer 3: ```suggestion
        final cBuilder = CBuilder.library(
          name: packageName,
          assetName: '$packageName.dart',
          sources: ['src/$packageName.c'],
        );
        await cBuilder.run(
```

          input: input,
          output: output,
          logger: Logger('')
            ..level = Level.ALL
            ..onRecord.listen((record) => print(record.message)),
        );
      });
    }
    ```

    The `build` function takes two arguments:

>     * `input`: (readonly) The input for the hook.

Reviewer 3: We don't use this `(readonly)` syntax anywhere else, so consider including that information directly in the sentence.

```suggestion
    * `input`: The read-only input for the hook.
```

      Includes information for the hook
      to produce the right asset type (for example,
      target OS, target architecture, output directory, and
      more). For details, see the [`BuildInput`][] class.

>     * `output`: (writeonly) The output for the hook. After the

Reviewer 3: ```suggestion
    * `output`: The write-only builder for the hook output. After the
```

      build hook reads the input, it produces an asset and
      then provides what it produced as the output.
      For details, see the [`BuildOutputBuilder`][] class.

[`BuildInput`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput-class.html
[`BuildOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/BuildOutputBuilder-class.html

### Automatically add assets {: #generate-assets }

> Assets are added indirectly when you use the `run`,
> `build`, or `test` command to invoke hooks on your package.

Reviewer 3: ```suggestion
Assets are added indirectly when you use the `run`,
`build`, or `test` commands to invoke hooks on your package.
```

The resulting assets are stored in the directory that you
specified in your hooks script. 

### Use assets {: #reference-assets }

Assets are the files that hooks create. Once an asset is

> created, you can reference it in your code and at runtime

Reviewer 3: ```suggestion
created, you can reference it in your code and at run time
```

with its asset ID ([`assetId`][]). Asset IDs are structured

> as `package:<package-name>/<asset-name>`. Build hooks may

Reviewer 3: https://developers.google.com/style/word-list#may

```suggestion
as `package:<package-name>/<asset-name>`. Build hooks can
```

only output assets in their own package. `CBuilder` in
the build hook in the previous example outputs the asset ID
`package:native_add_library/native_add_library.dart`, and is
based on the `packageName` and `assetName`.

The following example illustrates how to call the native

> C function `add` from `native_add_library.c`:

Reviewer 3: It seems the following code snippet is not necessarily calling the function, just initially binding to it. Perhaps change this to "how to bind" or add an example of calling it as well.

```dart title="my_package/lib/my_package.dart"
import 'dart:ffi';

@Native<Int32 Function(Int32, Int32)>()
external int add(int a, int b);
```

The asset ID in `@Native` is optional and defaults to the
library URI. In the previous example, this is
`package:native_add_library/native_add_library.dart`, which
is the same asset ID as output from the build hook. This

> enables Dart to connect an asset referenced at runtime to

Reviewer 3: ```suggestion
enables Dart to connect an asset referenced at run time to
```

| [`system_library`][]         | A Dart library using a native system library.                                     |
| [`use_dart_api`][]           | Project that uses the C API of the Dart VM.                                       |
 
{: .table .table-striped }

[`native_add_library`]: {{site.repo.dart.org}}/native/blob/main/pkgs/hooks/example/build/native_add_library
[`native_add_app`]: {{site.repo.dart.org}}/native/tree/main/pkgs/hooks/example/build/native_add_app
[`download_asset`]: {{site.repo.dart.org}}/native/tree/main/pkgs/hooks/example/build/download_asset
[`native_dynamic_linking`]: {{site.repo.dart.org}}/native/tree/main/pkgs/hooks/example/build/native_dynamic_linking
[`system_library`]: {{site.repo.dart.org}}/native/tree/main/pkgs/hooks/example/build/system_library
[`use_dart_api`]: {{site.repo.dart.org}}/native/tree/main/pkgs/hooks/example/build/use_dart_api

## More information

See the following for more information about hooks:

* [Hooks package][]
* [Hooks library reference][]
* [Build and bundle native code][]

[Hooks package]: {{site.pub-pkg}}/hooks
[Hooks library reference]: {{site.pub-api}}/hooks/latest/hooks/

> [Build and bundle native code]: https://dart.dev/interop/c-interop#build-hooks

Reviewer 3: ```suggestion
[Build and bundle native code]: /interop/c-interop#build-hooks
```

---
title: Hooks
description: Run custom build scripts.
---

:::note
Support for build hooks is in **preview** and
can be used on the `main` and `beta` channels.
:::

This guide explains what hooks are and how to use them
with a package.

## Introduction

You can currently use hooks to do things
such as compile or download native assets (code written in
other languages that are compiled into machine code), and
then call these assets from the Dart code of a package.

Hooks are Dart scripts placed in the `hook/` directory of
your Dart package. They have a predefined format for their
input and output, which allows the Dart SDK to:

1. Discover the hooks.
1. Execute the hooks with the necessary input.
1. Consume the output produced by the hooks.

Example project with a build hook:

```plaintext
📁  example_project                   // Project with hooks.
    📁  hook                          // Add hook scripts here.
        📄 build.dart

>     📁  lib                           // Us your assets here.

Reviewer: ```suggestion
    📁  lib                           // Use your assets here.
```

Link hooks are stored in the `hook/link.dart` directory
and create code assets and data assets in the
`build/` directory.

[`link`]: {{site.pub-api}}/hooks/0.19.5/hooks/link.html

## Assets

Assets are the files output from a hook that need to be
bundled in a Dart application. Assets can be accessed
at runtime from the Dart code. Currently, the Dart SDK can
use the `CodeAsset` and `DataAsset` asset types.

### CodeAsset type

A [`CodeAsset`][] represents a code asset. A code asset is
a dynamic library compiled from a language other than
Dart, such as C, C++, Rust, or Go. `CodeAsset` is part of
the `code_asset` package and are called with `dart:ffi` from
the Dart code at runtime using the `@Native` annotation.
A code asset can be used with build and link hooks.

[`CodeAsset`]: {{site.pub-api}}/code_assets/latest/code_assets/CodeAsset-class.html

### DataAsset type

A [`DataAsset`][] represents a data asset in bytes format.
It can be accessed by Dart at runtime and used in
Dart standalone packages. `DataAsset` is part of the
`data_asset` package and can be used with link hooks.

[`DataAsset`]: {{site.pub-api}}/data_assets/latest/data_assets/DataAsset-class.html

## Use a hook

> To setup a hook, use a hook to generate assets, and use

Reviewer 2: What did you want to say here?

To setup a hook use a hook sounds weird.

To add assets to your project, use ...

those assets in your Dart project, see the steps in the
following sections.

### Add dependencies

Open your package in a code editor, navigate to its
`pubspec.yaml`, and then add the following dependencies as
needed under `dependencies`:

* `hooks`
* `code_assets` (for build hooks and link hooks)
* `data_assets` (for link hooks)

You will also likely need to add a tool for your specific
task, such as `native_toolchain_c` for compiling C code.

:::note
You need to add the dependencies under `dependencies`, not
`dev_dependencies`. The hooks will be run by packages and
applications depending on your package, so the
Dart code needs to be part of the resolution in those
packages.
:::

> Example dependencies for a build hook:

Reviewer 2: It doesn't make sense to use link hooks without a build hook. The link hook is
meant to make assets from a build hook smaller.

Also, both build and link hooks can work with code assets. (And once data assets
support is added both hooks can work with data assets.)

          logger: Logger('')
            ..level = Level.ALL
            ..onRecord.listen((record) => print(record.message)),
        );
      });
    }
    ```

    * `input`: The input for `hook/build.dart`. This
      includes the target OS, target architecture,
      output directory, etc. For details, see
      the [`BuildInput`][] class.

    * `output`: The output for `hook/build.dart`. This
      includes the assets to be bundled, and dependencies.
      For details, see the [`BuildOutputBuilder`][] class.

[`BuildInput`]: {{site.pub-api}}/hooks/latest/hooks/BuildInput-class.html
[`BuildOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/BuildOutputBuilder-class.html

### Create a link hook script

If you want to use a link hook, follow these steps to
add one to a Dart script: 

1.  In your Dart project, create or open `hooks/link.dart`. 

1.  In the main method, add a `link` function from
    `package:hooks/hooks.dart` and construct a data asset.
    For example:

    ```dart title="hooks/link.dart" highlightLines=14
    import 'dart:convert';
    import 'dart:io';

    import 'package:data_assets/data_assets.dart';
    import 'package:hooks/hooks.dart';

>     import 'package:record_use/record_use.dart';

Reviewer 2: Record use is very experimental, we shouldn't show this yet.

Note that without record use you can't really use a link hook to shrink assets
yet. So maybe it would be better to leave link hooks out completely for now.

          ),
        );
      });
    }

    extension on LinkInput {
      RecordedUsages get usages {
        final usagesFile = recordedUsagesFile;
        final usagesContent = File.fromUri(usagesFile!).readAsStringSync();
        final usagesJson = jsonDecode(usagesContent) as Map<String, Object?>;
        final usages = RecordedUsages.fromJson(usagesJson);
        return usages;
      }
    }
    ```

    * `input`: The input for `hook/link.dart`. This
      includes the target OS, target architecture,
      output directory, etc. For details, see
      the [`LinkInput`][] class.

    * `output`: The output for `hook/link.dart`. This
      includes the assets to be bundled, and dependencies.
      For details, see the [`LinkOutputBuilder`][] class.

[`LinkInput`]: {{site.pub-api}}/hooks/latest/hooks/LinkInput-class.html
[`LinkOutputBuilder`]: {{site.pub-api}}/hooks/latest/hooks/LinkOutputBuilder-class.html

### Generate assets

Assets are generated when you invoke the Dart hooks in
the `hook/` directory. Your Dart hooks are invoked when when
you use the `run`, `build`, or `test` command on your
package.

### Reference assets

Assets are the files that hooks create in the

> `build/` directory. Once an asset has been created,

Reviewer 2: This is not true. It might be true for flutter, but not for Dart.

`dart build` creates a `build/` directory.

The assets are actually generated in `.dart_tool/build_runner/<...>/`, but I'm
not sure how much of that we want to tell users. They can see this from
inspecting `input.outputDirectoryShared`.

Author: Cool. I've updated the docs.
